### PR TITLE
Implemented strings for templates

### PR DESCRIPTION
**Pull request by UplinkCoder**

### CODE REVIEW

**Comment by schancel** [comment]
@UplinkCoder Garrr.  Come on man, clean this up!  This is a good PR, but it needs a bit of polish on the error handling.

**Comment by UplinkCoder** [comment]
I just leave the dtemplate changes out of it.
The string-mangling is non-controversal.

**Comment by deadalnix** [comment]
@schancel taht is not necessary. >> is well defined in D, unlike in C, and so the compiler will ensure it does the right thing.

File: src/d/semantic/mangler.d [hunk 1]
```diff
@@ -210,7 +210,24 @@ struct ValueMangler {
...
-	
+
+	string visit(StringLiteral s) {
+		import util.hex;
+
+		auto ret = "a";
+		auto str = s.value;
+		auto len = str.length;
+		
+		ret.reserve(len * 2 +  (len < 100) ? 4 : 8);
```
Comment (deadalnix) [inline src/d/semantic/mangler.d]: I can't find the spec for reserve. It is an absolute value there or how much you want to add ?

Comment (UplinkCoder) [inline src/d/semantic/mangler.d]: pretty sure is absolute

Comment (schancel) [inline src/d/semantic/mangler.d]: Also, just do `len * 2 + 8`

Comment (deadalnix) [inline src/d/semantic/mangler.d]: Yes, make it 8 and be done.

File: src/d/semantic/mangler.d [hunk 2]
```diff
@@ -210,7 +210,24 @@ struct ValueMangler {
...
+		ret ~= to!string(len);
+		ret ~= '_';
+		foreach(immutable ubyte c; str) {
```
Comment (deadalnix) [inline src/d/semantic/mangler.d]: unless you take by ref, immutable is not needed here.

File: src/util/hex.d [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿module util.hex;
+
+const (char[2]) byte2hex(const ubyte b) pure {
```
Comment (deadalnix) [inline src/util/hex.d]: Same, const is not needed. It is pass by value anyway.

**Comment by schancel** [comment]
Just write a unit tests and this should be good.   We need to implement `.mangleof`to be able to write a test case.

**Comment by deadalnix** [comment]
Yes it is fairly easy to add unitests block on that one. Also, I'm not sure it is worth to add a new module for that function. Just put it as private in mangler.

**Comment by UplinkCoder** [comment]
@schancel added small test.

**Pull request closed without merging**